pallets/chainlink-price-manager/src/lib.rs

```diff
@@ -0,0 +1,260 @@
+				.ok_or(OffchainErr::FailReceivingOraclePrice)?
+				.latest_round();
+			min_round_id = min_round_id.min(feed_result);
+		}
+		Ok(min_round_id)
+	}
+
+	// Print prices to node debug log
+	fn print_prices() {
+		for currency in CurrencyId::get_enabled_tokens_in_protocol(UnderlyingAsset) {
+			if let Some(price) = Self::get_underlying_price(currency) {
+				log::debug!("{:?} price is {:?}", currency, price);
+			} else {
+				log::warn!("Can't receive price for {:?}", currency);
+			}
+		}
+	}
+
+	fn _offchain_worker(now: T::BlockNumber) -> Result<(), OffchainErr> {
+		// TODO implement extrinsic to set initiate round period instead of hardcoded 3
+		let bn: T::BlockNumber = (3_u32).into();
+		if (now % bn).is_zero() {
+			// TODO should we get latest_round for each pool and produce event pair, or enough take
+			// only min round id?
+			let new_round_id = Self::get_min_round_id()?.saturating_add(1);
+			// Currently feed id is not play any role. See TODO above
+			let feed_id = Self::get_feed_id(ETH).ok_or(OffchainErr::ChainlinkFeedNotExists)?;
+			let call = Call::<T>::initiate_new_round(feed_id, new_round_id);
+			log::debug!("New round_id: {:?}", new_round_id);
+
+			if SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call.into()).is_err() {
+				log::info!(
+					target: "ChainlinkPriceManager offchain worker",
+					"Initiate a new round is faled",
+				);
+			}
+		}
+		Self::print_prices();
+		Ok(())
+	}
+
+	// TODO This is temporary function. We need tom move this function as method to
+	// privitives/src/currency.rs. Also, add distingiush between chainlink provider and minterest
+	fn convert_to_description(currency_id: CurrencyId) -> &'static [u8] {
+		match currency_id {
+			ETH => b"MIN-ETH",
+			DOT => b"MIN-DOT",
+			KSM => b"MIN-KSM",
+			BTC => b"MIN-BTC",
+			// This must be gone after implementing strict CurrencyId types
+			_ => b"Non-existent-feed",
+		}
+	}
+
+	/// Looks for appropriate feed config with description and returns FeedId
+	pub fn get_feed_id(currency_id: CurrencyId) -> Option<T::FeedId> {
+		Some(
+			<pallet_chainlink_feed::Feeds<T> as IterableStorageMap<
+				T::FeedId,
+				pallet_chainlink_feed::FeedConfigOf<T>,
+			>>::iter()
+			.find(|(_, v)| v.description == Self::convert_to_description(currency_id))?
+			.0,
+		)
+	}
+}
+
+impl<T: Config> PricesManager<CurrencyId> for Pallet<T>
+where
+	u128: From<<T as pallet_chainlink_feed::Config>::Value>,
+{
+	fn get_underlying_price(currency_id: CurrencyId) -> Option<Price> {
+		// TODO check is feeding enabled
+		let feed_id = Self::get_feed_id(currency_id)?;
+		let feed_result = <ChainlinkFeedPallet<T>>::feed(feed_id)?;
+		// TODO handle updated_at
+		let RoundData { answer, .. } = feed_result.latest_data();
+
+		// There is an issue that pallet-chainlink-feed can return Some(0)
+		// if feed was created but submit() extrinsic wasn't called
+		if answer.is_zero() {
+			return None;
+		}
+
+		Some(Price::from_inner(answer.into()))
```

Comment:
The associated type Value implement `trait BaseArithmetic`: https://github.com/smartcontractkit/chainlink-polkadot/blob/master/pallet-chainlink-feed/src/lib.rs#L298.
`traits BaseArithmetic` depends on traits `TryFrom<u128> + TryInto<u128> + UniqueSaturatedFrom<u128> + UniqueSaturatedInto<u128>`: https://substrate.dev/rustdocs/v3.0.0-monthly-2021-05/sp_arithmetic/traits/trait.BaseArithmetic.html
So I think there is no need to add a constraint:
`where  u128: From<<T as pallet_chainlink_feed::Config>::Value>,`
and you can write like this:
 ```suggestion
		Some(Price::from_inner(answer.try_into().ok()?))
```
or
```suggestion
		Some(Price::from_inner(answer.unique_saturated_into()))
```

Reply:
Nice suggestion :+1:  Improved